src/common/mod.rs

```diff
@@ -130,6 +130,7 @@ where
 
             while self.session.wants_write() {
                 match self.write_io(cx) {
+                    Poll::Ready(Ok(0)) => return Poll::Ready(Err(io::ErrorKind::WriteZero.into())),
```

Comment:
Thoughts about doing this inside `write_io`, rather than each caller of it?

That could maybe look like:

```patch
diff --git a/src/common/mod.rs b/src/common/mod.rs
index 5dc6a05..3ba96cc 100644
--- a/src/common/mod.rs
+++ b/src/common/mod.rs
@@ -387,12 +387,18 @@ impl<T: Unpin> SyncWriteAdapter<'_, '_, T> {
 impl<T: AsyncWrite + Unpin> Write for SyncWriteAdapter<'_, '_, T> {
     #[inline]
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        self.poll_with(|io, cx| io.poll_write(cx, buf))
+        match self.poll_with(|io, cx| io.poll_write(cx, buf)) {
+            Ok(0) => Err(io::ErrorKind::WriteZero.into()),
+            other => other,
+        }
     }
 
     #[inline]
     fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
-        self.poll_with(|io, cx| io.poll_write_vectored(cx, bufs))
+        match self.poll_with(|io, cx| io.poll_write_vectored(cx, bufs)) {
+            Ok(0) => Err(io::ErrorKind::WriteZero.into()),
+            other => other,
+        }
     }
 
     fn flush(&mut self) -> io::Result<()> {

```

Reply:
Currently `return 0` is handled as pending in `poll_write` (see https://github.com/rustls/tokio-rustls/blob/v/0.26.0/src/common/mod.rs#L273), i would like to keep this behavior.

because `poll_write` will first write buf to session, we need to return consumed length, otherwise the user cannot safely call `poll_write` again. `flush` and `shutdown` do not need to do this, because there is no user input, so i prefer to return error.

Reply:
Makes sense!